1.21.0 breaks the build for iOS

@muxi: I believe that may concern you.

I build gRPC and its deps from source, using CMake, with [this simple repo (branch: ios)](https://github.com/JonasVautherin/grpc-build/tree/ios). Note that I apply [a patch](https://github.com/JonasVautherin/grpc-build/blob/ios/grpc/grpc_ios.patch) to gRPC, that I think is not relevant to this issue (I will open a PR with that patch, btw).

Building 1.20.0 is fine, using the following command:

```
cmake -DCMAKE_TOOLCHAIN_FILE=tools/ios.toolchain.cmake -Bbuild/ios -S.
```

Now, building 1.21.0 fails with the following error

```
Scanning dependencies of target grpc++
[ 77%] Linking C executable grpc_create_jwt.app/grpc_create_jwt
ld: warning: -headerpad_max_install_names is ignored when used with -bitcode_bundle (Xcode setting ENABLE_BITCODE=YES)
[ 77%] Building CXX object CMakeFiles/grpc++.dir/src/cpp/client/insecure_credentials.cc.o
Undefined symbols for architecture armv7:
  "operator delete(void*)", referenced from:
      CFStreamHandle::Unref(char const*, int, char const*) in libgrpc.a(cfstream_handle.cc.o)
  "operator new(unsigned long)", referenced from:
      CFStreamHandle::CreateStreamHandle(__CFReadStream*, __CFWriteStream*) in libgrpc.a(cfstream_handle.cc.o)
ld: symbol(s) not found for architecture armv7
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[5]: *** [grpc_verify_jwt.app/grpc_verify_jwt] Error 1
make[4]: *** [CMakeFiles/grpc_verify_jwt.dir/all] Error 2
make[4]: *** Waiting for unfinished jobs....
Undefined symbols for architecture armv7:
  "operator delete(void*)", referenced from:
      CFStreamHandle::Unref(char const*, int, char const*) in libgrpc.a(cfstream_handle.cc.o)
  "operator new(unsigned long)", referenced from:
      CFStreamHandle::CreateStreamHandle(__CFReadStream*, __CFWriteStream*) in libgrpc.a(cfstream_handle.cc.o)
```

Am I doing something wrong, or is that a bug in gRPC?

This can be easily reproduced by:

1. Cloning my repo (note: you want the "ios" branch):
```
    git clone -b ios https://github.com/jonasvautherin/grpc-build
```

2. Building 1.20.0 to convince yourself it works:
```
    cmake -DCMAKE_TOOLCHAIN_FILE=tools/ios.toolchain.cmake -Bbuild/ios_1.20.0 -S.
```

3. Changing [grpc/CMakeLists.txt](https://github.com/JonasVautherin/grpc-build/blob/ios/grpc/CMakeLists.txt#L40) to point to 1.21.0 instead of 1.20.0, and build that to see the error:
```
    cmake -DCMAKE_TOOLCHAIN_FILE=tools/ios.toolchain.cmake -Bbuild/ios_1.21.0 -S.
```

Note: there is a [bug in CMake](https://gitlab.kitware.com/cmake/cmake/issues/19282) for the current version distributed on Homebrew. You can either use the `pip`-provided cmake or the CMake nightly build.

--- a/src/core/lib/iomgr/cfstream_handle.cc
+++ b/src/core/lib/iomgr/cfstream_handle.cc
@@ -19,6 +19,7 @@
 #include <grpc/support/port_platform.h>
 
 #include "src/core/lib/iomgr/port.h"
+#include "src/core/lib/gprpp/memory.h"
 
 #ifdef GRPC_CFSTREAM
 #import <CoreFoundation/CoreFoundation.h>
@@ -47,8 +48,7 @@ void CFStreamHandle::Release(void* info) {
 
 CFStreamHandle* CFStreamHandle::CreateStreamHandle(
     CFReadStreamRef read_stream, CFWriteStreamRef write_stream) {
-  CFStreamHandle* handle = static_cast<CFStreamHandle*>(gpr_malloc(sizeof(CFStreamHandle)));
-  return new (handle) CFStreamHandle(read_stream, write_stream);
+  return grpc_core::New<CFStreamHandle>(read_stream, write_stream);
 }
 
 void CFStreamHandle::ReadCallback(CFReadStreamRef stream,
@@ -189,8 +189,7 @@ void CFStreamHandle::Unref(const char* file, int line, const char* reason) {
             reason, val, val - 1);
   }
   if (gpr_unref(&refcount_)) {
-    this->~CFStreamHandle();
-    gpr_free(this);
+    grpc_core::Delete<CFStreamHandle>(this);
   }
 }
 

--- a/src/core/lib/iomgr/cfstream_handle.cc
+++ b/src/core/lib/iomgr/cfstream_handle.cc
@@ -47,7 +47,8 @@ void CFStreamHandle::Release(void* info) {
 
 CFStreamHandle* CFStreamHandle::CreateStreamHandle(
     CFReadStreamRef read_stream, CFWriteStreamRef write_stream) {
-  return new CFStreamHandle(read_stream, write_stream);
+  CFStreamHandle* handle = static_cast<CFStreamHandle*>(gpr_malloc(sizeof(CFStreamHandle)));
+  return new (handle) CFStreamHandle(read_stream, write_stream);
 }
 
 void CFStreamHandle::ReadCallback(CFReadStreamRef stream,
@@ -188,7 +189,8 @@ void CFStreamHandle::Unref(const char* file, int line, const char* reason) {
             reason, val, val - 1);
   }
   if (gpr_unref(&refcount_)) {
-    delete this;
+    this->~CFStreamHandle();
+    gpr_free(this);
   }
 }
 

--- a/src/core/lib/iomgr/cfstream_handle.cc
+++ b/src/core/lib/iomgr/cfstream_handle.cc
@@ -18,8 +18,8 @@
 
 #include <grpc/support/port_platform.h>
 
-#include "src/core/lib/iomgr/port.h"
 #include "src/core/lib/gprpp/memory.h"
+#include "src/core/lib/iomgr/port.h"
 
 #ifdef GRPC_CFSTREAM
 #import <CoreFoundation/CoreFoundation.h>

--- a/src/core/lib/iomgr/cfstream_handle.h
+++ b/src/core/lib/iomgr/cfstream_handle.h
@@ -29,6 +29,7 @@
 #ifdef GRPC_CFSTREAM
 #import <CoreFoundation/CoreFoundation.h>
 
+#include "src/core/lib/gprpp/memory.h"
 #include "src/core/lib/iomgr/closure.h"
 #include "src/core/lib/iomgr/lockfree_event.h"
 
@@ -65,6 +66,9 @@ class CFStreamHandle final {
   dispatch_queue_t dispatch_queue_;
 
   gpr_refcount refcount_;
+
+  GRPC_ALLOW_CLASS_TO_USE_NON_PUBLIC_NEW
+  GRPC_ALLOW_CLASS_TO_USE_NON_PUBLIC_DELETE
 };
 
 #ifdef DEBUG

